### PR TITLE
add get_uri function to protected operation #8

### LINKED ISSUE #8: Protected resource URI generation should be more flexible.
Currently the protected resource decorator assumes that the URI of the resource it's protecting can be generated based on a simple template and the resource object's id. While this works most of the time, there will be cases where that's too simplistic (See OnroerendErfgoed/adviezen#210).

Eg. when a certain resource has different URI's based on a subtype of a resource. Eg, when we have a resource shapes that has two subtypes: circles and squares. The URI's in use might be `shapes/<id>`, but they might also be `circles/<id>` and `squares/<id>` based on the subtype.

This might also happen because we have a uri like `https://id.erfgoed.net/actoren/501/adressen/15`. In this case a template with just an id is insufficient.

Maybe we can implement this with a callback function that is passed the resource it's protecting? But maybe we currently don't have the resource until after we've executed the view?

/cc @BartSaelen

### LINKED ISSUE #9: Error message when a backend application can't be reached
When a service is not reachable for the uriregistry.

AS IS:
HTTPInternalServerError: Urireferencer: Something went wrong while retrieving references of the uri https://id.erfgoed.net/besluiten/1

TO BE:
HTTP 503 Service Unavailable: Urireferencer: Something went wrong while retrieving references of the uri https://id.erfgoed.net/test/1. Application A not reachable.

### PR DESCRIPTION
**Pull request by cahytinne**
fix #8 
fix #9

### CODE REVIEW

**Comment by koenedaele** [comment]
Comments from #8 :+1: 

>So, in this case if we have both `circle` and `square` resources, they should have different View classes with different functions?
>
>It's a good fix for this issue, but it does not help with #7, but maybe we can tackle that one later.
>
>Also, can we change the function name to something like urireferencer_get_uri()? To reduce the possibilities of name clashes.
>
>I think it also breaks BC right now. Can we have a default version of urireferencer_get_uri that reads the uri_template attribute and combines that with the id from the matchdict? I think that would maintain BC?

@BartSaelen Could you provide some input as well please?

**Comment by koenedaele** [comment]
Any idea if we should use the abc library (think that's what I used in skosprovider) or zope interface (https://pypi.python.org/pypi/zope.interface). That last one seems to be rather heavily used in pyramid land?

**Comment by BartSaelen** [comment]
A would prefer ABC, but if you want Zope, it's ok

**Comment by koenedaele** [comment]
I'm ok with abc. I've never really investigated zope.interface and was wondering if we were missing something relevant.

**Comment by cahytinne** [comment]
Alternative solution:

- `urireferencer_get_uri` as abstract method of the AbstractReferencer with the request as input variable.

That way, the link with the View class is not that strong.

**Comment by cahytinne** [comment]
Are the changes ok?

It will break the current version, so the user will be forced to implement a `get_uri` function to the existing`referencer` using the request.

**Comment by koenedaele** [comment]
Can we implement a default function in the abstract referencer that must receive the request, but can also receive an optional keyword uri_template. If this keyword is set, the function searches for an id in the matchdict (throws some kind of error if no id is present) and renders the uri like the advice function does it now. If no uri_template is passed, we can raise a not implemented error.

I think this allows maintaining BC. We could make it a deprecated option that will be removed in eg. 1.0.0.

File: pyramid_urireferencer/referencer.py [hunk 1]
```diff
@@ -44,12 +56,12 @@ def is_referenced(self, uri):
         """
 
 
+@six.add_metaclass(abc.ABCMeta)
 class Referencer(AbstractReferencer):
     """
     This is an implementation of the :class:`AbstractReferencer` that adds a
-    generic :meth:`is_referenced` method and a plain :meth:`references` method..
+    generic :meth:`is_referenced` method and plains methods: :meth:`references` and :meth:`get_uri`
```
Comment (koenedaele) [inline pyramid_urireferencer/referencer.py]: plains should be plain

File: pyramid_urireferencer/protected_resources.py [hunk 2]
```diff
@@ -33,16 +32,6 @@ def protected_operation(fn):
     def advice(parent_object, *args, **kw):
         referencer = pyramid_urireferencer.get_referencer(parent_object.request.registry)
         uri = referencer.get_uri(parent_object.request)
-        if uri is None:
```
Comment (koenedaele) [inline pyramid_urireferencer/protected_resources.py]: Not sure where this code went. But we should not raise a 501 when the urireferencing fails. For the client this is an HTTP 500 Internal Server Error and nothing more. We should have some good logging for the server admin though.

**Pull request merged**